# 배경
- Sepsis와 항생제
    - 맨 처음 항생제 처방 시점
    - 항생체 첫 처방 후 다른 약으로 변경(내가 항생제를 잘 처방하는게 맞나? 바꿔야 하나?)
    - 타임 윈도우를 좀더 세밀하게 나누어서
        - intime
        - outtime
        - antibiotic_time_poe
        - suspected_of_infection_poe



# 문제 종류
- binary classification: 항생제를 처방받았는가
- [TBD] multi-label classification: 어떤 항생제를 처방받았는가

- 유지, 끊기, 바꾸기를 예측
- sepsis3_cohort의 주요 피처들을 좀더 세밀한 타임 윈도우로 추출해야함
     - antibiotic_time_poe: 첫 처방인지 마지막 처방인지
     - vent 등 스코어나 체온 등을 세밀하게 뽑을 수 있는지 

# 실험
- 코호트

<img src="images/pnp_fig1.png">

- 피처 전처리/엔지니어링
    - [TBD]
- 모델
    - LR
    - RF
    - GB
    - (R)NN

# 교차검증
- 5-folds CV

# 평가
- AUROC
- variable importance: 어떤 변수가 다음 처방에 영향을 미쳤을까?


# 예측값의 활용
- 처방 실수의 예방

In [4]:
%matplotlib inline

import matplotlib.pyplot as plt
import seaborn as sns; sns.set(rc={'figure.figsize':(15,15)})
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
engine = create_engine('postgresql://postgres:mimic@127.0.0.1:5555/mimic')

In [50]:
sql = """
select a.subject_id, s.* 
from sepsis3_cohort s 
    left join icustays a 
    on s.hadm_id=a.hadm_id and s.icustay_id=a.icustay_id
where s.excluded = 0
"""
label = pd.read_sql(sql, engine)
label.head().T

0  \
subject_id                                            165   
hadm_id                                            170252   
icustay_id                                         247247   
intime                                2170-10-03 17:17:53   
outtime                               2170-10-04 16:24:15   
age                                               86.4824   
gender                                                  M   
ethnicity                          BLACK/AFRICAN AMERICAN   
first_service                                         MED   
dbsource                                       metavision   
suspected_of_infection_poe                              1   
suspected_infection_time_poe          2170-10-03 11:20:00   
suspected_infection_time_poe_days                 0.24853   
specimen_poe                                BLOOD CULTURE   
positiveculture_poe                                     0   
antibiotic_time_poe                   2170-10-03 00:00:00   
exclusion_secondarystay                                 0   
exclusion_nonadult                                      0   
exclusion_csurg                                         0   
exclusion_carevue                                       0   
exclusion_early_suspicion                               0   
exclusion_late_suspicion                                0   
exclusion_bad_data                                      0   
excluded                                                0   

                                                        1  \
subject_id                                            266   
hadm_id                                            186251   
icustay_id                                         293876   
intime                                2168-07-10 08:02:23   
outtime                               2168-07-11 17:40:38   
age                                               77.5623   
gender                                                  F   
ethnicity                          BLACK/AFRICAN AMERICAN   
first_service                                        NMED   
dbsource                                       metavision   
suspected_of_infection_poe                              1   
suspected_infection_time_poe          2168-07-10 09:26:00   
suspected_infection_time_poe_days              -0.0580671   
specimen_poe                                  MRSA SCREEN   
positiveculture_poe                                     0   
antibiotic_time_poe                   2168-07-12 00:00:00   
exclusion_secondarystay                                 0   
exclusion_nonadult                                      0   
exclusion_csurg                                         0   
exclusion_carevue                                       0   
exclusion_early_suspicion                               0   
exclusion_late_suspicion                                0   
exclusion_bad_data                                      0   
excluded                                                0   

                                                     2                    3  \
subject_id                                         422                  671   
hadm_id                                         117029               126769   
icustay_id                                      299666               246119   
intime                             2173-04-03 11:36:29  2195-12-31 04:45:35   
outtime                            2173-04-08 14:20:31  2196-01-07 17:35:50   
age                                            57.6467              47.7629   
gender                                               M                    M   
ethnicity                                        WHITE                WHITE   
first_service                                     CMED                NSURG   
dbsource                                    metavision           metavision   
suspected_of_infection_poe                           0                    1   
suspected_infection_time_poe                       NaT  

In [29]:
label.groupby('excluded').size()

excluded
0    11791
1    49741
dtype: int64

In [30]:
label.subject_id.nunique()

46476

In [31]:
label[label.excluded==0].subject_id.nunique()

11791

In [32]:
label[label.excluded==1].subject_id.nunique()

36839

In [33]:
sql = """
select count(distinct subject_id)
from icustays
"""
pd.read_sql(sql, engine)

count
0  46476

- icu의 총 환자 46476명 중에 sepsis3 조건에 맞는 icustay는 11791 건
    - 해당 조건의 고유 환자수도 11791건으로 icustays:subejct_id = 1:1
    - 즉, 각 환자가 한번만 icu에 입원했고, sepsis 진단을 받았다. 여러번 입원하고 진단받은 사례는 없다.
    - 따라서, 환자 수준의 모델링만 가능, 방문 수준의 모델링은 불가능

<hr>

In [40]:
label[(label.excluded==0) & (label.antibiotic_time_poe.notnull())]['icustay_id'].count()

7588

In [41]:
label[(label.excluded==0) & (label.antibiotic_time_poe.isnull())]['icustay_id'].count()

4203

In [43]:
label[(label.excluded==0) & (label.antibiotic_time_poe.notnull())].subject_id.nunique()

7588

In [44]:
label[(label.excluded==0) & (label.antibiotic_time_poe.isnull())].subject_id.nunique()

4203

- icu의 총 환자 46476명 중에 sepsis3 조건에 맞는 icustay는 11791 건
    - 항생제 처방을 받은 환자수 7588, 건수 7588, 
    - 항생제 처방을 받지 않은 환자수 4203, 건수는 4203

<hr>

In [47]:
sql = """
select i.subject_id, s.*, a.antibiotic_time_poe
from sepsis3 s 
    join sepsis3_cohort a 
        on s.icustay_id=a.icustay_id
    join icustays i
        on s.icustay_id=i.icustay_id
where a.excluded = 0
"""
features = pd.read_sql(sql, engine)

In [48]:
features.head().T

0                    1  \
subject_id                                       61691                41710   
icustay_id                                      200021               200028   
hadm_id                                         109307               181955   
excluded                                             0                    0   
intime                             2114-12-26 19:45:12  2133-10-29 17:13:50   
outtime                            2114-12-27 22:46:28  2133-11-01 14:55:14   
dbsource                                    metavision           metavision   
suspected_infection_time_poe                       NaT                  NaT   
suspected_infection_time_poe_days                  NaN                  NaN   
specimen_poe                                      None                 None   
positiveculture_poe                                NaN                  NaN   
antibiotic_time_poe                                NaT                  NaT   
blood_culture_time                 2114-12-27 15:00:00  2133-10-29 19:27:00   
blood_culture_positive                               0                    0   
age                                            60.8501              64.8666   
gender                                               M                    M   
is_male                                              1                    1   
ethnicity                                        WHITE                WHITE   
race_white                                           1                    1   
race_black                                           0                    0   
race_hispanic                                        0                    0   
race_other                                           0                    0   
metastatic_cancer                                    0                    0   
diabetes                                             0                    1   
height                                             NaN               167.64   
weight                                            82.6                   84   
bmi                                                NaN              29.8899   
first_service                                     NMED                 CMED   
hospital_expire_flag                                 0                    0   
thirtyday_expire_flag                                0                    0   
icu_los                                         1.1259               2.9038   
hosp_los                                       1.94861              3.20417   
sepsis_angus                                         0                    0   
sepsis_martin                                        0                    0   
sepsis_explicit                                      0                    0   
septic_shock_explicit                                0                    0   
severe_sepsis_explicit                               0                    0   
sepsis_nqf                                           0                    0   
sepsis_cdc                                           0                    0   
sepsis_cdc_simple                                    0                    0   
elixhauser_hospital                                  4                    8   
vent                                                 1                    1   
sofa                                                 4                   10   
lods                                                 7                    6   
sirs                                                 3                    3   
qsofa                                                3                    2   
qsofa_sysbp_score                                    1                    1   
qsofa_gcs_score                                      1                    0   
qsofa_resprate_score                                 1                    1   
antibiotic_time_poe                                NaT                  NaT   

                       

# TBD

- 위의 피처 전처리를 어떻게 할것인가
- 추가로 붙일 피처가 있을까
- sepsis 인데 항생제 처방을 안 받는게 어떤 의미인가

# Reference 

[1] Sepsis-3 in MIMIC-III, https://github.com/alistairewj/sepsis3-mimic